File: CSharp/DomWalker/DomApi.cs
using HtmlWalker;
using System.Collections.Generic;
using System.IO;
using System.Text;

namespace DomWalker
{
    public abstract class DomApi<D>
    {
        internal protected DomApi() { }

        internal protected abstract object DocumentNode(D dom);

        internal protected abstract object HtmlElement(D dom);

        internal protected abstract string Html(D dom);

        internal protected abstract object CreateElement(D dom, string name);

        internal protected abstract object CreateTextNode(D dom, string text);

        internal protected abstract object CreateEntityNode(D dom, string entity);

        internal protected abstract IList<object> GetContent(D dom, ITag adapter);

        internal protected abstract void Load(D dom, string path);

        internal protected abstract void Load(D dom, TextReader reader);

        internal protected abstract void LoadHtml(D dom, string html);

        internal protected abstract void Save(D dom, string path, Encoding encoding);

        internal protected abstract void Save(D dom, TextWriter writer);
    }
}

File: Java/htmlwalker/src/main/java/domwalker/DomApi.java
package domwalker;

import java.io.Writer;
import java.util.List;

import htmlwalker.ITag;
import htmlwalker.exception.HtmlWalkerException;

public abstract class DomApi<D>
{
	protected DomApi() { }
	
    public abstract Object documentNode(D dom);
    
    public abstract Object htmlElement(D dom);
    
    public abstract String html(D dom);

    public abstract Object createElement(D dom, String name);

    public abstract Object createTextNode(D dom, String text);

    public abstract Object createEntityNode(D dom, String entity);

    public abstract List<Object> getContent(D dom, ITag adapter);
    
    public abstract D load(D dom, String path) throws HtmlWalkerException;

    public abstract D loadHtml(D dom, String html) throws HtmlWalkerException;    	    

    public abstract void save(D dom, String path, String encoding) throws HtmlWalkerException;

    public abstract void save(D dom, String path) throws HtmlWalkerException;

    public abstract void save(D dom, Writer writer) throws HtmlWalkerException;
}
